fix(cli): skip empty image-verify responses for ImageValidatingPolicy tests

Denish3436 opened the pull request:
…exclusions

## Explanation
This PR prevents the CLI test command from emitting redundant “Excluded” entries when running tests against an ImageValidatingPolicy (IVPOL) that produces no image-verification rules. 
<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue
closes #12988 
<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)

My PR contains new or altered behavior to Kyverno. 
- [ ] I have sent the draft PR to add or update [the documentation](https://github.com/kyverno/website) and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->

## What type of PR is this
/kind bug


<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

MariamFahmy98 reviewed (changes requested):
Thanks for your contribution!

First of all, the bug exists in the new ValidatingPolicy type so your fix won't actually address the issue.
The reason behind having an additional entry with an `Exclude` as a reason is because the following condition is true and its code is executed which it shouldn't:
https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/commands/test/output.go#L290-L306

To debug it, you need first to check the responses returned from the engine.

Assuming that we have the following manifests:
`policy.yaml`:
```
apiVersion: policies.kyverno.io/v1alpha1
kind: ValidatingPolicy
metadata:
  name: disallow-host-path
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   [""]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["pods"]
  validations:
    - expression: "!has(object.spec.volumes) || object.spec.volumes.all(volume, !has(volume.hostPath))"
      message: "HostPath volumes are forbidden. The field spec.volumes[*].hostPath must be unset."
```
`deployment1.yaml`:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: baddeployment01
spec:
  replicas: 1
  selector:
    matchLabels:
      app: app
  template:
    metadata:
      labels:
        app: app
    spec:
      containers:
      - name: container01
        image: dummyimagename
        volumeMounts:
          - name: udev
            mountPath: /data
      volumes:
      - name: udev
        hostPath:
          path: /etc/udev
```
`deployment2.yaml`:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gooddeployment01
spec:
  replicas: 1
  selector:
    matchLabels:
      app: app
  template:
    metadata:
      labels:
        app: app
    spec:
      containers:
      - name: container01
        image: dummyimagename
```
`kyverno-test.yaml`:
```
apiVersion: cli.kyverno.io/v1alpha1
kind: Test
metadata:
  name: kyverno-test
policies:
- policy.yaml
resources:
- deployment1.yaml
- deployment2.yaml
results:
- isValidatingPolicy: true
  kind: Deployment
  policy: disallow-host-path
  resources:
  - baddeployment01
  result: fail
- isValidatingPolicy: true
  kind: Deployment
  policy: disallow-host-path
  resources:
  - gooddeployment01
  result: pass
```
If you run the `test` command, the output will be as follows:
```
./cmd/cli/kubectl-kyverno/kubectl-kyverno test test/cli/test-validating-policy/disallow-host-path
Loading test  ( test/cli/test-validating-policy/disallow-host-path/kyverno-test.yaml ) ...
  Loading values/variables ...
  Loading policies ...
  Loading resources ...
  Loading exceptions ...
  Applying 1 policy to 2 resources with 0 exceptions ...
  Checking results ...

│────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
│ ID │ POLICY             │ RULE │ RESOURCE                                    │ RESULT │ REASON   │
│────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
│ 1  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
│ 2  │ disallow-host-path │      │ apps/v1/Deployment/default/baddeployment01  │ Pass   │ Ok       │
│ 3  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
│ 4  │ disallow-host-path │      │ apps/v1/Deployment/default/gooddeployment01 │ Pass   │ Ok       │
│────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│


Test Summary: 4 tests passed and 0 tests failed
```
Currently, when processing Deployments, we observe three engine responses. This is because we have three distinct validations: one for Pods, one for resource types like Deployment/StatefulSet,..etc, and one for CronJobs. For a Deployment, only one validation matches, leaving the other two engine responses with empty policy results.

The main problem is the creation of engine responses that contain no policy response. The policy processor (it calls the engine and processes the responses) needs a check: it should only proceed to create an engine response if the rules array is not empty. Otherwise, it should skip the creation of that response.

So we need to add this check here before creating the engine response.
https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/processor/policy_processor.go#L315-L323

I assume that the same issue exists for ImageValidatingPolicies too so you have to check it too.

Let me know if you have any questions.

Denish3436 commented:
> Thanks for your contribution!
> 
> First of all, the bug exists in the new ValidatingPolicy type so your fix won't actually address the issue. The reason behind having an additional entry with an `Exclude` as a reason is because the following condition is true and its code is executed which it shouldn't:
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/commands/test/output.go#L290-L306
> 
> To debug it, you need first to check the responses returned from the engine.
> 
> Assuming that we have the following manifests: `policy.yaml`:
> 
> ```
> apiVersion: policies.kyverno.io/v1alpha1
> kind: ValidatingPolicy
> metadata:
>   name: disallow-host-path
> spec:
>   matchConstraints:
>     resourceRules:
>     - apiGroups:   [""]
>       apiVersions: ["v1"]
>       operations:  ["CREATE", "UPDATE"]
>       resources:   ["pods"]
>   validations:
>     - expression: "!has(object.spec.volumes) || object.spec.volumes.all(volume, !has(volume.hostPath))"
>       message: "HostPath volumes are forbidden. The field spec.volumes[*].hostPath must be unset."
> ```
> 
> `deployment1.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: baddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
>         volumeMounts:
>           - name: udev
>             mountPath: /data
>       volumes:
>       - name: udev
>         hostPath:
>           path: /etc/udev
> ```
> 
> `deployment2.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: gooddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
> ```
> 
> `kyverno-test.yaml`:
> 
> ```
> apiVersion: cli.kyverno.io/v1alpha1
> kind: Test
> metadata:
>   name: kyverno-test
> policies:
> - policy.yaml
> resources:
> - deployment1.yaml
> - deployment2.yaml
> results:
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - baddeployment01
>   result: fail
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - gooddeployment01
>   result: pass
> ```
> 
> If you run the `test` command, the output will be as follows:
> 
> ```
> ./cmd/cli/kubectl-kyverno/kubectl-kyverno test test/cli/test-validating-policy/disallow-host-path
> Loading test  ( test/cli/test-validating-policy/disallow-host-path/kyverno-test.yaml ) ...
>   Loading values/variables ...
>   Loading policies ...
>   Loading resources ...
>   Loading exceptions ...
>   Applying 1 policy to 2 resources with 0 exceptions ...
>   Checking results ...
> 
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ ID │ POLICY             │ RULE │ RESOURCE                                    │ RESULT │ REASON   │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ 1  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 2  │ disallow-host-path │      │ apps/v1/Deployment/default/baddeployment01  │ Pass   │ Ok       │
> │ 3  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 4  │ disallow-host-path │      │ apps/v1/Deployment/default/gooddeployment01 │ Pass   │ Ok       │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> 
> 
> Test Summary: 4 tests passed and 0 tests failed
> ```
> 
> Currently, when processing Deployments, we observe three engine responses. This is because we have three distinct validations: one for Pods, one for resource types like Deployment/StatefulSet,..etc, and one for CronJobs. For a Deployment, only one validation matches, leaving the other two engine responses with empty policy results.
> 
> The main problem is the creation of engine responses that contain no policy response. The policy processor (it calls the engine and processes the responses) needs a check: it should only proceed to create an engine response if the rules array is not empty. Otherwise, it should skip the creation of that response.
> 
> So we need to add this check here before creating the engine response.
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/processor/policy_processor.go#L315-L323
> 
> I assume that the same issue exists for ImageValidatingPolicies too so you have to check it too.
> 
> Let me know if you have any questions.

Hey @MariamFahmy98 Mariam, thanks again for the feedback! I’ve updated the PR with these changes:



> Thanks for your contribution!
> 
> First of all, the bug exists in the new ValidatingPolicy type so your fix won't actually address the issue. The reason behind having an additional entry with an `Exclude` as a reason is because the following condition is true and its code is executed which it shouldn't:
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/commands/test/output.go#L290-L306
> 
> To debug it, you need first to check the responses returned from the engine.
> 
> Assuming that we have the following manifests: `policy.yaml`:
> 
> ```
> apiVersion: policies.kyverno.io/v1alpha1
> kind: ValidatingPolicy
> metadata:
>   name: disallow-host-path
> spec:
>   matchConstraints:
>     resourceRules:
>     - apiGroups:   [""]
>       apiVersions: ["v1"]
>       operations:  ["CREATE", "UPDATE"]
>       resources:   ["pods"]
>   validations:
>     - expression: "!has(object.spec.volumes) || object.spec.volumes.all(volume, !has(volume.hostPath))"
>       message: "HostPath volumes are forbidden. The field spec.volumes[*].hostPath must be unset."
> ```
> 
> `deployment1.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: baddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
>         volumeMounts:
>           - name: udev
>             mountPath: /data
>       volumes:
>       - name: udev
>         hostPath:
>           path: /etc/udev
> ```
> 
> `deployment2.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: gooddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
> ```
> 
> `kyverno-test.yaml`:
> 
> ```
> apiVersion: cli.kyverno.io/v1alpha1
> kind: Test
> metadata:
>   name: kyverno-test
> policies:
> - policy.yaml
> resources:
> - deployment1.yaml
> - deployment2.yaml
> results:
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - baddeployment01
>   result: fail
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - gooddeployment01
>   result: pass
> ```
> 
> If you run the `test` command, the output will be as follows:
> 
> ```
> ./cmd/cli/kubectl-kyverno/kubectl-kyverno test test/cli/test-validating-policy/disallow-host-path
> Loading test  ( test/cli/test-validating-policy/disallow-host-path/kyverno-test.yaml ) ...
>   Loading values/variables ...
>   Loading policies ...
>   Loading resources ...
>   Loading exceptions ...
>   Applying 1 policy to 2 resources with 0 exceptions ...
>   Checking results ...
> 
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ ID │ POLICY             │ RULE │ RESOURCE                                    │ RESULT │ REASON   │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ 1  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 2  │ disallow-host-path │      │ apps/v1/Deployment/default/baddeployment01  │ Pass   │ Ok       │
> │ 3  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 4  │ disallow-host-path │      │ apps/v1/Deployment/default/gooddeployment01 │ Pass   │ Ok       │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> 
> 
> Test Summary: 4 tests passed and 0 tests failed
> ```
> 
> Currently, when processing Deployments, we observe three engine responses. This is because we have three distinct validations: one for Pods, one for resource types like Deployment/StatefulSet,..etc, and one for CronJobs. For a Deployment, only one validation matches, leaving the other two engine responses with empty policy results.
> 
> The main problem is the creation of engine responses that contain no policy response. The policy processor (it calls the engine and processes the responses) needs a check: it should only proceed to create an engine response if the rules array is not empty. Otherwise, it should skip the creation of that response.
> 
> So we need to add this check here before creating the engine response.
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/processor/policy_processor.go#L315-L323
> 
> I assume that the same issue exists for ImageValidatingPolicies too so you have to check it too.
> 
> Let me know if you have any questions.

Hey @MariamFahmy98, thanks for the feedback! I’ve updated the PR. Could you take another look and see if it's all good now?

Denish3436 commented:
> Thanks for your contribution!
> 
> First of all, the bug exists in the new ValidatingPolicy type so your fix won't actually address the issue. The reason behind having an additional entry with an `Exclude` as a reason is because the following condition is true and its code is executed which it shouldn't:
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/commands/test/output.go#L290-L306
> 
> To debug it, you need first to check the responses returned from the engine.
> 
> Assuming that we have the following manifests: `policy.yaml`:
> 
> ```
> apiVersion: policies.kyverno.io/v1alpha1
> kind: ValidatingPolicy
> metadata:
>   name: disallow-host-path
> spec:
>   matchConstraints:
>     resourceRules:
>     - apiGroups:   [""]
>       apiVersions: ["v1"]
>       operations:  ["CREATE", "UPDATE"]
>       resources:   ["pods"]
>   validations:
>     - expression: "!has(object.spec.volumes) || object.spec.volumes.all(volume, !has(volume.hostPath))"
>       message: "HostPath volumes are forbidden. The field spec.volumes[*].hostPath must be unset."
> ```
> 
> `deployment1.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: baddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
>         volumeMounts:
>           - name: udev
>             mountPath: /data
>       volumes:
>       - name: udev
>         hostPath:
>           path: /etc/udev
> ```
> 
> `deployment2.yaml`:
> 
> ```
> apiVersion: apps/v1
> kind: Deployment
> metadata:
>   name: gooddeployment01
> spec:
>   replicas: 1
>   selector:
>     matchLabels:
>       app: app
>   template:
>     metadata:
>       labels:
>         app: app
>     spec:
>       containers:
>       - name: container01
>         image: dummyimagename
> ```
> 
> `kyverno-test.yaml`:
> 
> ```
> apiVersion: cli.kyverno.io/v1alpha1
> kind: Test
> metadata:
>   name: kyverno-test
> policies:
> - policy.yaml
> resources:
> - deployment1.yaml
> - deployment2.yaml
> results:
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - baddeployment01
>   result: fail
> - isValidatingPolicy: true
>   kind: Deployment
>   policy: disallow-host-path
>   resources:
>   - gooddeployment01
>   result: pass
> ```
> 
> If you run the `test` command, the output will be as follows:
> 
> ```
> ./cmd/cli/kubectl-kyverno/kubectl-kyverno test test/cli/test-validating-policy/disallow-host-path
> Loading test  ( test/cli/test-validating-policy/disallow-host-path/kyverno-test.yaml ) ...
>   Loading values/variables ...
>   Loading policies ...
>   Loading resources ...
>   Loading exceptions ...
>   Applying 1 policy to 2 resources with 0 exceptions ...
>   Checking results ...
> 
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ ID │ POLICY             │ RULE │ RESOURCE                                    │ RESULT │ REASON   │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> │ 1  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 2  │ disallow-host-path │      │ apps/v1/Deployment/default/baddeployment01  │ Pass   │ Ok       │
> │ 3  │ disallow-host-path │      │ apps/Deployment/v1                          │ Pass   │ Excluded │
> │ 4  │ disallow-host-path │      │ apps/v1/Deployment/default/gooddeployment01 │ Pass   │ Ok       │
> │────│────────────────────│──────│─────────────────────────────────────────────│────────│──────────│
> 
> 
> Test Summary: 4 tests passed and 0 tests failed
> ```
> 
> Currently, when processing Deployments, we observe three engine responses. This is because we have three distinct validations: one for Pods, one for resource types like Deployment/StatefulSet,..etc, and one for CronJobs. For a Deployment, only one validation matches, leaving the other two engine responses with empty policy results.
> 
> The main problem is the creation of engine responses that contain no policy response. The policy processor (it calls the engine and processes the responses) needs a check: it should only proceed to create an engine response if the rules array is not empty. Otherwise, it should skip the creation of that response.
> 
> So we need to add this check here before creating the engine response.
> 
> https://github.com/kyverno/kyverno/blob/ae665a58cd635cd4bb7d33b63a988cf2d8f704c2/cmd/cli/kubectl-kyverno/processor/policy_processor.go#L315-L323
> 
> I assume that the same issue exists for ImageValidatingPolicies too so you have to check it too.
> 
> Let me know if you have any questions.

Hey @MariamFahmy98 @Mohdcode , Thanks for the feedback. I have updated the PR. Could you take another look and see if it’s all good?

Mohdcode commented on cmd/cli/kubectl-kyverno/processor/policy_processor.go:
```diff
@@ -167,6 +167,10 @@ func (p *PolicyProcessor) ApplyPoliciesOnResource() ([]engineapi.EngineResponse,
 			return responses, err
 		}
 		verifyImageResponse, verifiedImageData := eng.VerifyAndPatchImages(context.TODO(), policyContext)
+		if len(verifyImageResponse.PolicyResponse.Rules) == 0 {
```
i think this is for `verifyImage` rule and not for`ImageValidatingPolicy`.

MariamFahmy98 replied:
As mentioned by @Mohdcode, this change should be for ImageValidatingPolicies. You need to check that `rules` are not empty before proceeding to create an engine response.

Mohdcode commented:
thanks @Denish3436 I added a comment, let me know if you see same error while reproducing the issue with `IVPOL`.

MariamFahmy98 commented on pkg/autogen/v1/autogen.go:
Can you revert the changes here as it doesn't fix the bug?

MariamFahmy98 commented on cmd/cli/kubectl-kyverno/processor/policy_processor.go:
```diff
@@ -312,15 +312,18 @@ func (p *PolicyProcessor) ApplyPoliciesOnResource() ([]engineapi.EngineResponse,
-					PolicyResponse: engineapi.PolicyResponse{
-						Rules: r.Rules,
-					},
+				if len(r.Rules) > 0 {
+					response := engineapi.EngineResponse{
+						Resource: *reps.Resource,
+						PolicyResponse: engineapi.PolicyResponse{
+							Rules: r.Rules,
+						},
+					}
```
I suggest we first check if the `rules` is empty. If it is, we should skip creating a response. For example:
```
if len(r.Rules) == 0 {
   continue
}
```

MariamFahmy98 reviewed (changes requested):
We need to add the same check for ImageValidatingPolicies too.
https://github.com/Mohdcode/kyverno/blob/441cb4b555f16388e28e3f99854f73a5b50c676f/cmd/cli/kubectl-kyverno/commands/test/test.go#L450-L455

MariamFahmy98 commented:
Some CLI tests are failing. Could you please check them?

Mohdcode commented:
@Denish3436 let me know if you need any help or i can pick it.
